thorlcr/slave/slavmain.cpp

```diff
@@ -153,21 +157,70 @@ class CJobListener : public CSimpleInterface
             return true;
         }
     } excptHandler;
+
 public:
     CJobListener() : excptHandler(*this)
     {
         stopped = true;
+        processPerSlave = globals->getPropBool("@processPerSlave");
+        slavesPerNode = globals->getPropInt("@slavesPerNode", 1);
+        mpServers.append(* getMPServer());
+        if (!processPerSlave)
+        {
+            unsigned port = getMachinePortBase();
+            // will need generating into thor.xml, currently only in setvars
+            unsigned localThorPortInc = globals->getPropInt("@localThorPortInc", 200);
+            for (unsigned sc=1; sc<slavesPerNode; sc++)
+            {
+                port += localThorPortInc;
+                mpServers.append(*startNewMPServer(port));
+            }
+        }
     }
     ~CJobListener()
     {
+        for (unsigned sc=1; sc<slavesPerNode; sc++)
+            mpServers.item(sc).stop();
+        mpServers.kill();
         stop();
     }
     void stop()
     {
-        queryClusterComm().cancel(0, masterSlaveMpTag);
+        queryNodeComm().cancel(0, masterSlaveMpTag);
     }
     virtual void main()
     {
+        if (!processPerSlave)
+        {
+            class CVerifyThread : public CInterface, implements IThreaded
+            {
+                CThreaded threaded;
+                CJobListener &jobListener;
+                unsigned channel;
+            public:
+                CVerifyThread(CJobListener &_jobListener, unsigned _channel)
+                    : jobListener(_jobListener), channel(_channel), threaded("CVerifyThread", this)
+                {
+                    start();
+                }
+                ~CVerifyThread() { join(); }
+                void start() { threaded.start(); }
+                void join() { threaded.join(); }
+                virtual void main()
+                {
+                    Owned<ICommunicator> comm = jobListener.mpServers.item(channel).createCommunicator(&queryClusterGroup());
+                    PROGLOG("verifying mp connection to rest of slaves (from channel=%d)", channel);
+                    if (!comm->verifyAll())
+                        ERRLOG("Failed to connect to reset of slaves");
+                    else
+                        PROGLOG("verified mp connection to rest of slaves");
+                }
+            };
+            CIArrayOf<CInterface> verifyThreads;
+            for (unsigned c=0; c<slavesPerNode; c++)
```

Comment:
No, because it has already done a verifyAll on the nodeCommunicator by this stage.
slave 0 above share the same IMPServer as the nodeCommunicator on each node, so will already be connected.